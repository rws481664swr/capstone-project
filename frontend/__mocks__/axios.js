import {BASE_URL} from "../src/config";

axios.get.mockImplementation((url) => {
    let result = url.slice(BASE_URL.length);
    if(result[result.length-1] === '/') result = result.slice(0, result.length-1);
    switch (result){
        case '/courses':
            return Promise.resolve({data: coursesList})
        case '/users':
            return Promise.resolve({data:usersList})
        case '/auth/token':
            return Promise.resolve({data:'token'})
    }
    return Promise.resolve({data: []})

})
const axios = {
    get: jest.fn().mockResolvedValue({
        data: {}
    }),
    post: jest.fn().mockResolvedValue({
        data: {}
    }),
    put: jest.fn().mockResolvedValue({
        data: {}
    }),
    delete: jest.fn().mockResolvedValue({
        data: {}
    })
}
export default axios