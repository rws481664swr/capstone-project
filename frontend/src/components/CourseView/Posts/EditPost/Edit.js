import useForm from "../../../../hooks/useForm";
import LabeledInput, {LabeledTextBox} from "../../../General/LabeledInput";
import {useDispatch} from "react-redux";
import useAxios from "../../../../hooks/useAxios";
import './EditPost.css'
import Button from "../../../General/Button";


const Edit = ({  post, setEditMode, setPost}) => {
    const [form, onChange, resetForm] = useForm({
        title: post.title, content: post.content
    })
    const dispatch = useDispatch()
    const {put} = useAxios()
    const updateForm = async () => {
        try {
            await put(`posts/${post._id}`, '', form)
            dispatch({type: 'UPDATE', payload: {...post, ...form}})
            setPost({...post, ...form})
            setEditMode(false)
            resetForm()
        } catch (e) {
            console.error(e)
        }
    }
    return <form onSubmit={updateForm}>
            <div>
                <LabeledInput
                    id={'edit-title'}
                    label={'Title'}
                    value={form.title}
                    onChange={onChange}
                    className={'LabeledInput'}
                    name={'title'}
                />

                <LabeledTextBox
                    id={'edit-content'}
                    label={'Content'}
                    value={form.body}
                    onChange={onChange}
                    className={'LabeledTextBox'}
                    name={'content'}
                    children={form.content}
                />
                <Button type="submit">Save</Button>
            </div>
    </form>
}
export default Edit