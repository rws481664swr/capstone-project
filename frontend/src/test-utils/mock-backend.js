import {BASE_URL} from "../config";

export const extractKeysAsList= (obj)=>
    Object.keys(obj).map(key=>obj[key])
        .reduce((acc, val)=>[...acc,val],[])
export const MOCK_TOKEN = {current: 'TOKEN'}

export const spliceUrl = (url) => {
    const arr = BASE_URL.split('')
    const toReturn = url.split('')
    while (arr.length){
        arr.shift()
        toReturn.shift()
    }
    return toReturn.join('')

}