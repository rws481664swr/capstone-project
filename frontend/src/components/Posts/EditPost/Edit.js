import useForm from "../../../hooks/useForm";
import LabeledInput, {LabeledTextBox} from "../../General/LabeledInput/LabeledInput";
import {useDispatch, useSelector} from "react-redux";
import useAxios from "../../../hooks/useAxios";
import './EditPost.css'
import Button from "../../General/Button/Button";
import {UPDATE} from "../../../state/actions/actions";
import {useCallback} from "react";


const Edit = ({  post, setEditMode, setPost}) => {
    const [form, onChange, resetForm] = useForm({
        title: post.title, content: post.content
    })
    const store = useSelector(e=>e)

    const dispatch = useDispatch()
    const {put} = useAxios()
    const updateForm =async (e)  => {
        e.preventDefault()

        try {
            await put(`posts/${post._id}`, '', form)
            console.log("UPDATED")

            dispatch({type: UPDATE , payload: {...post, ...form,id:post._id}})
            setPost({...post, ...form})
            console.log("UPDATED 2")
            setEditMode(false)
            resetForm()
        } catch (e) {
            console.error(e)
        }
    }
    return <form onSubmit={updateForm}>
            <div>
                <LabeledInput
                    id={'edit-title'}
                    label={'Title'}
                    value={form.title}
                    onChange={onChange}
                    className={'LabeledInput'}
                    name={'title'}
                />

                <LabeledTextBox
                    id={'edit-content'}
                    label={'Content'}
                    value={form.content}
                    onChange={onChange}
                    className={'LabeledTextBox'}
                    name={'content'}
                 />
                <Button  onClick={(e)=> updateForm(e)} >Save</Button>
            </div>
    </form>
}
export default Edit