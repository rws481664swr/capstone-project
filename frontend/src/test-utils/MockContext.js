import GlobalContext from "../state/contexts/GlobalContext";
import {MOCK_TOKEN} from "./mock-backend";
import {MemoryRouter} from "react-router-dom";

const MockContext = ({children, isAdmin = false, ...mockValue}) => {
    return (
        <MemoryRouter>
            <GlobalContext.Provider value={{
                token: MOCK_TOKEN,
                _id: 'id',
                username: 'username',
                role: isAdmin ? "ADMIN" : "STUDENT",
                loggedIn: true,
                timestamp: new Date(),
                ...mockValue
            }}>
                {children}
            </GlobalContext.Provider>
        </MemoryRouter>
    )
}

export default MockContext