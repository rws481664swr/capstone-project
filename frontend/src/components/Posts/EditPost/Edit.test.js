import {fireEvent, render} from "@testing-library/react";
import Edit from "./Edit";
import MockContext from "../../../test-utils/MockContext";
import {useDispatch} from "react-redux";
import {act} from "react-dom/test-utils";
import axios from "axios";

jest.mock('axios', () => ({
    get: jest.fn(),
    post: jest.fn(),
    put: jest.fn(),
    delete: jest.fn()
}))
const mockDispatch = jest.fn()
jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn(),
    useDispatch: jest.fn(() => () => {
    })
}))
it('should render Edit', async () => {

    const post = {
        _id: '123',
        title: 'test',
        content: 'test-content'
    }



    useDispatch.mockImplementation(() => (mockDispatch))

    // configureStore({reducer})
    const {asFragment, getByText} =
        render(<MockContext>
            <Edit post={post} setEditMode={jest.fn()} setPost={jest.fn()} editMode={true}/>)
        </MockContext>)
    getByText('test-content')
 expect(asFragment()).toMatchSnapshot()
})
it('should edit a post', async () => {
    const editing = {isEditing: true}

    const setEditMode = jest.fn((mode) => {
        editing.isEditing = mode
    })
    const setPost = jest.fn((post) => {
        editing.post = post
    })
    const post = {
        _id: '123',
        title: 'test',
        content: 'test-content'
    }
    axios.put.mockImplementation(async (url, body) => {
        console.log(url)
        console.log(body)
        expect(body).toEqual({title: 'test', content: 'changed'})
        expect(url).toContain('123')
        console.log("AXIOSPUT")
        return {
            data: body
        }
    })
    // const reducer = (state,action)=>{
    //     if(action.type[0]==='@'){
    //         return state
    //     }
    //
    // }


    useDispatch.mockImplementation(() => (mockDispatch))

    // configureStore({reducer})
    const {getByLabelText, getByText} =
        render(<MockContext>
            <Edit post={post} setEditMode={setEditMode} setPost={setPost} editMode={editing.isEditing}/>)
        </MockContext>)
    const title = getByLabelText('Title')
    const content = getByLabelText('Content')
    expect(title).toBeInTheDocument()
    expect(content).toBeInTheDocument()
    fireEvent.change(title, {target: {value: 'test'}})
    fireEvent.change(content, {target: {value: 'changed'}})
    await act(() => fireEvent.click(getByText('Save')))
    expect(axios.put).toHaveBeenCalled()
    expect(setEditMode).toHaveBeenCalledWith(false)
    expect(editing.isEditing).toBe(false)
    expect(setPost).toHaveBeenCalledWith({...post, title: 'test', content: 'changed'})
    expect(editing.post).toEqual({...post, title: 'test', content: 'changed'})


})