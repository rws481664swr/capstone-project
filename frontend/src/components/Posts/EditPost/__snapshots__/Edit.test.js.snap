// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit 1`] = `
<DocumentFragment>
  <form>
    <div>
      <div
        class="LabeledInput-div LabeledInput"
      >
        <label
          class="LabeledInput-input-label "
          for="edit-title"
        >
          Title
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          id="edit-title"
          name="title"
          type="text"
          value="test"
        />
      </div>
      <div
        class="LabeledInput-div LabeledTextBox"
      >
        <label
          class="LabeledInput-label "
          for="edit-content"
        >
          Content
        </label>
        <textarea
          class="LabeledInput-textarea "
          id="edit-content"
          name="content"
        >
          test-content
        </textarea>
      </div>
      <button
        class="undefined Button"
      >
        Save
      </button>
    </div>
  </form>
  )
</DocumentFragment>
`;
